### PR TITLE
fix indirect broadcast

### LINKED ISSUE #142250: Embedding forward performance analysis
### 🐛 Describe the bug

Embedding operation maps indices into a dense vector space, typically used to look up embeddings for discrete input tokens in tasks like natural language processing.

## **Reproduce tests**

```
% python run.py --op embedding  --mode fwd --num-inputs 1 --input-id 11 --precision fp32 --metrics latency,speedup --cudagraph
  0%|                                                                                                                                                         | 0/1 [00:00<?, ?it/s]INFO:tritonbench.utils.triton_op:Took 178.71ms to get benchmark function for torch_embedding
INFO:tritonbench.utils.triton_op:Took 178.65ms to get benchmark function for liger_embedding
INFO:tritonbench.utils.triton_op:Took 822.35ms to get benchmark function for inductor_embedding
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.54s/it]
         (B, T, D, V)    torch_embedding-latency    liger_embedding-speedup    liger_embedding-latency    inductor_embedding-speedup    inductor_embedding-latency
---------------------  -------------------------  -------------------------  -------------------------  ----------------------------  ----------------------------
(8, 2048, 4096, 8192)                   0.472864                    2.50288                   0.188928                        1.9795                       0.23888
```

## Report Analysis

For input 11, the input dimensions are \``(B, T, D, V) = (8, 2048, 4096, 8192)`\`, so 

```
embeddings.shape = (V, D) = (8192, 4096)
indices.shape = (B, T) = (8, 2048)
output.shape = (B, T, D) = (8, 2048, 4096)
```

For FP32, the theoretical optimal write bytes should be 

sizeof(output) \= 8\*2048\*4096\*4/1024/1024 \= 256MB

The theoretical optimal read bytes should be 

sizeof(indices \+ embeddings) \= (8\*2048  \+  \*8192\*4096) \* 4/1024/1024 MB= 128.0625MB 

### Inductor compiled triton kernel

The following is the compiled triton kernel and its wrapper code.

In the `grid=grid(67108864)` function, it will compute the final grid size.   
[https://github.com/pytorch/pytorch/blob/a84779040049377aec7b62e37becb7327950541e/torch/\_inductor/runtime/triton\_heuristics.py\#L1962-L2004](https://github.com/pytorch/pytorch/blob/a84779040049377aec7b62e37becb7327950541e/torch/_inductor/runtime/triton_heuristics.py#L1962-L2004)

Kernel triton\_poi\_fused\_embedding\_0’s final grid size \= (65536, 1,1) while its XBLOCK= 1024\.  
So, theoretical memory write request per warp is

65536\*1024/32/4 \=524,288 

32 threads per warp, 4 FP32 per STG.E.128

It matches the memory write request in ncu reports. 
image

However, for memory read, the indices load is compiled to the folllowing triton line.

| x1 \= (xindex // 4096\) tmp0 \= tl.load(in\_ptr0 \+ (x1), None, eviction\_policy='evict\_last') |
| :---- |

Because each thread will execute this load, So the total read requests for indices are also 524,288. So the total number of load requests are 1048576\. This matches the memory table in ncu report. 
image

I also confirmed with a modified [mem\_trace tool](https://github.com/FindHao/nvbit_test2) in nvbit. 

`MEMTRACE: Instruction execution counts for kernel triton_poi_fused_embedding_0 in context 0x55ed1dd95330:`  
 `Instruction: STG.E.128            Count: 524288`  
 `Instruction: LDG.E.128            Count: 524288`  
 `Instruction: LDG.E.EL.64          Count: 524288`

LDG.E.128 for embedding loads, LDG.E.EL.64 for indices loads.

### Liger implementation

[https://github.com/linkedin/Liger-Kernel/blob/main/src/liger\_kernel/ops/experimental/embedding.py\#L81-L95](https://github.com/linkedin/Liger-Kernel/blob/main/src/liger_kernel/ops/experimental/embedding.py#L81-L95)  
For liger version, it converts shapes to make it compatible with two-dimension blocks (128, 128), and use gridsize \= (128,32), which is much smaller than inductor compiled version. Triton also automatically place the smaller indices to shared memory. Thus, the liger version saves a lot of global memory loads. 

### Error logs

_No response_

### Versions

main branch

### PR DESCRIPTION
**Pull request by FindHao**
Fixes #142250
```bash
python run.py --op embedding --mode fwd --precision fp32 --metrics latency,speedup --csv
```
The performance data is collected as below. 
| (B, T, D, V)            | latency |       |                     |                    | speedup |                     |                    |
| ----------------------- | ------- | ----- | ------------------- | ------------------ | ------- | ------------------- | ------------------ |
|                         | torch   | liger | inductor-before-fix | inductor-after-fix | liger   | inductor-before-fix | inductor-after-fix |
| (32, 512, 768, 1024)    | 0.086   | 0.029 | 0.029               | 0.029              | 2.954   | 3.014               | 3.023              |
| (32, 512, 768, 2048)    | 0.088   | 0.032 | 0.032               | 0.032              | 2.783   | 2.743               | 2.753              |
| (32, 512, 768, 4096)    | 0.091   | 0.034 | 0.037               | 0.035              | 2.642   | **2.444**           | 2.575              |
| (32, 512, 768, 8192)    | 0.095   | 0.039 | 0.042               | 0.040              | 2.423   | **2.236**           | 2.387              |
| (32, 512, 768, 16384)   | 0.099   | 0.045 | 0.047               | 0.045              | 2.224   | **2.117**           | 2.213              |
| (32, 512, 768, 32768)   | 0.102   | 0.048 | 0.050               | 0.049              | 2.136   | **2.059**           | 2.103              |
| (32, 512, 768, 65536)   | 0.105   | 0.050 | 0.052               | 0.051              | 2.092   | 2.027               | 2.053              |
| (32, 512, 768, 131072)  | 0.107   | 0.052 | 0.052               | 0.052              | 2.062   | 2.053               | 2.046              |
| (8, 2048, 4096, 1024)   | 0.431   | 0.144 | 0.173               | 0.137              | 3.002   | **2.495**           | 3.145              |
| (8, 2048, 4096, 2048)   | 0.459   | 0.153 | 0.204               | 0.149              | 2.995   | **2.249**           | 3.075              |
| (8, 2048, 4096, 4096)   | 0.484   | 0.168 | 0.226               | 0.166              | 2.883   | **2.139**           | 2.906              |
| (8, 2048, 4096, 8192)   | 0.495   | 0.189 | 0.238               | 0.190              | 2.615   | **2.084**           | 2.608              |
| (8, 2048, 4096, 16384)  | 0.506   | 0.215 | 0.246               | 0.216              | 2.355   | **2.063**           | 2.349              |
| (8, 2048, 4096, 32768)  | 0.513   | 0.235 | 0.249               | 0.234              | 2.184   | **2.063**           | 2.195              |
| (8, 2048, 4096, 65536)  | 0.516   | 0.245 | 0.250               | 0.244              | 2.110   | 2.066               | 2.117              |
| (8, 2048, 4096, 131072) | 0.516   | 0.252 | 0.251               | 0.250              | 2.051   | 2.059               | 2.067              |

When building memory dependencies for a scheduler node, we check if this memory dependency is an indirect broadcast (i.e., indirect load indices from a tensor via one dimension and broadcasting them to another dimension to index the current memory dependency). If so, we record the dimensions for the pairs of (indirect load dimension, broadcast dimension). When deciding whether tiling needs to be applied, this information is used to generate a new loop order for the dimensions. If the loop order differs from the current one, we apply the new loop order and apply tilings if possible.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

**Comment by shunting314** [comment]
Can you add some test cases (e.g. embedding lookup) and show some perf numbers with the test case

**Comment by FindHao** [comment]
> Can you add some test cases (e.g. embedding lookup) and show some perf numbers with the test case

What do you mean show perf numbers? like adding the embedding op, then test the performance number before and after this PR and list the numbers here or in the comments? 

There is already a [`test_indirect_load_broadcast`](https://github.com/pytorch/pytorch/blob/c644f4c5febf71927ff5feb57b6631dfdfb49c6c/test/inductor/test_torchinductor.py#L7498), 
I'm thinking of adding a code check to check the compiled kernel using the right dimension to indexing indices for correctness check.

**Comment by FindHao** [comment]
The unit test is added https://github.com/pytorch/pytorch/pull/145992/commits/255f5c3992e9c606aa4a2e7eca803c1063d35395 I'll update the PR description for more details including performance numbers

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -7513,6 +7513,26 @@ def fn(in_ptr0, in_ptr1, in_ptr2):
...
+        )
+        # Check if the indices tensor are accessed via x dimension and the tiling works
+        for string in [
+            "xnumel = 16384",
+            "xoffset = tl.program_id(0) * XBLOCK",
+            "xindex = xoffset + tl.arange(0, XBLOCK)[:, None]",
+            "x1 = xindex",
+            "tmp0 = tl.load(in_ptr0 + (x1), None,",
+        ]:
+            self.assertTrue(string, code)
```
Comment (shunting314) [inline test/inductor/test_torchinductor.py:7534]: Your test should always pass since 'string' is 'True'

**Comment by shunting314** [comment]
I'd like to see the PR work as expected to improve perf (at least for micro-benchmark). Your test is a good setup. We should add code to benchmark the perf. Here is an example: https://github.com/pytorch/pytorch/blob/8af79b7ec816f5c73536a806aa4c7ea1f7bd3867/test/inductor/test_torchinductor.py#L12530

File: test/inductor/test_torchinductor.py (hunk 1 above)
Comment (FindHao) [reply test/inductor/test_torchinductor.py]: Thanks for catching it! will fix it in next commit!

**Comment by FindHao** [comment]
> I'd like to see the PR work as expected to improve perf (at least for micro-benchmark). Your test is a good setup. We should add code to benchmark the perf. Here is an example:
> 
> https://github.com/pytorch/pytorch/blob/8af79b7ec816f5c73536a806aa4c7ea1f7bd3867/test/inductor/test_torchinductor.py#L12530

Since the perf comparisons are between the versions before and after this PR, do you think printing out the execution time is good enough just like the case you pointed?

**Comment by shunting314** [comment]
> Since the perf comparisons are between the versions before and after this PR, do you think printing out the execution time is good enough just like the case you pointed?
yea

**Comment by FindHao** [comment]
> I'd like to see the PR work as expected to improve perf (at least for micro-benchmark). Your test is a good setup. We should add code to benchmark the perf. Here is an example:
> 
> https://github.com/pytorch/pytorch/blob/8af79b7ec816f5c73536a806aa4c7ea1f7bd3867/test/inductor/test_torchinductor.py#L12530

add a perf run https://github.com/pytorch/pytorch/pull/145992/commits/8dddda821b69c76e97c4367745f4042158c5f577 
Before this PR, the output is 
```
torch_ms=0.426
inductor_ms=0.205
speedup=2.07x
```
After this PR, the output is 
```
torch_ms=0.426
inductor_ms=0.171
speedup=2.49x
```

File: test/inductor/test_torchinductor.py [hunk 2]
```diff
@@ -7513,6 +7513,37 @@ def fn(in_ptr0, in_ptr1, in_ptr2):
             ),
         )
 
+    def test_indirect_broadcast_embedding(self):
+        B, T, D, V = (8, 2048, 4096, 2048)
+        op = nn.Embedding(V, D).to(self.device).to(torch.float32)
+        shared_weight = op.weight.data
+        tmp_op = nn.Embedding(V, D).to(self.device).to(torch.float32)
+        tmp_op.weight.data.copy_(shared_weight)
```
Comment (shunting314) [inline test/inductor/test_torchinductor.py:7521]: why you need to make a copy of the embedding op?

File: torch/_inductor/codegen/simd.py [hunk 3]
```diff
@@ -2034,9 +2034,37 @@ def convert_tiling_to_3d(
 
         if len(ranked_tilings) > 1:
             perf_hint_log.info("possibly bad tiling: %s", ranked_tilings)
+        # save iter vars in indirect broadcast memory reads
+        indirect_broadcast_pairs: list[tuple[sympy.Symbol, sympy.Symbol]] = []
+        if len(node_schedule) == 1:
```
Comment (shunting314) [inline torch/_inductor/codegen/simd.py:2039]: Can this restriction be loosen? What if there are fusion happening?

File: torch/_inductor/codegen/simd.py [hunk 4]
```diff
@@ -2034,9 +2034,37 @@ def convert_tiling_to_3d(
...
+            for read in node_schedule[0].read_writes.reads:
+                if (
+                    isinstance(read, MemoryDep)
+                    and hasattr(read, "indirect_broadcast")
+                    and read.indirect_broadcast
```
Comment (shunting314) [inline torch/_inductor/codegen/simd.py:2044]: getattr(read, "indirect_broadcast", None)

File: torch/_inductor/codegen/simd.py [hunk 5]
```diff
@@ -2034,9 +2034,37 @@ def convert_tiling_to_3d(
...
+                ):
+                    indirect_broadcast_pairs.append(read.indirect_broadcast)
+
+            if indirect_broadcast_pairs:
```
Comment (shunting314) [inline torch/_inductor/codegen/simd.py:2048]: Add comment how this part works especially how you decide the new loop order

File: torch/_inductor/dependencies.py [hunk 6]
```diff
@@ -69,12 +69,19 @@ class MemoryDep(Dep):
     var_names: tuple[sympy.Symbol, ...]
     size: tuple[sympy.Expr, ...]
     mode: Optional[str] = None
+    # store the replacements for the index when normalizing
+    replacement: tuple[tuple[sympy.Symbol, sympy.Expr], ...] = ()
+    # store the indirect broadcast dimensions
+    indirect_broadcast: Union[tuple[()], tuple[sympy.Symbol, sympy.Symbol]] = ()
```
Comment (shunting314) [inline torch/_inductor/dependencies.py:75]: Comment the meaning of the first and second symbol

File: torch/_inductor/dependencies.py [hunk 7]
```diff
@@ -664,6 +731,82 @@ def extract_loop_body_with_args(
...
+            if any(
+                x is None for x in (additive_symbol, coefficient, multiplied_symbol)
+            ):
+                continue
+            # check if the coefficient is the stride
+            buffer = None
+            if read.name in V.graph.name_to_buffer:
+                buffer = V.graph.name_to_buffer[read.name]  # type: ignore[assignment]
+            elif read.name in V.graph.graph_inputs:
+                buffer = V.graph.graph_inputs[read.name]  # type: ignore[assignment]
```
Comment (shunting314) [inline torch/_inductor/dependencies.py:756]: Is there an API for this?

**Comment by shunting314** [comment]
Can `indirect_broadcast` be computed on the fly rather than storing it as new fields for MemoryDep?

File: test/inductor/test_torchinductor.py (hunk 2 above)
Comment (FindHao) [reply test/inductor/test_torchinductor.py:7521]: If compile the previous `op`, the inductor_ms is always 0.09 for both before and after this fix. I don't know why it happens and can't reproduce it in an [independent script](https://github.com/FindHao/ml_scripts/blob/main/inductor/cases/embedding/run_embedding.py). any ideas?

File: torch/_inductor/codegen/simd.py (hunk 3 above)
Comment (FindHao) [reply torch/_inductor/codegen/simd.py:2039]: not sure about this. let me try, see if there are ci errors and see if I can write a test.

File: test/inductor/test_torchinductor.py (hunk 2 above)
Comment (shunting314) [reply test/inductor/test_torchinductor.py:7521]: self.common will cast the model to low precision. You can pass in check_lowp=False and make the test cleaner. But not really a big deal

Comment (FindHao) [reply test/inductor/test_torchinductor.py]: `check_lowp=False` works! fixed in https://github.com/pytorch/pytorch/pull/145992/commits/59e131647fe1bdc9a3c53b8643421629a44ff8fc

File: torch/_inductor/codegen/simd.py (hunk 3 above)
Comment (FindHao) [reply torch/_inductor/codegen/simd.py]: loose the restriction in https://github.com/pytorch/pytorch/pull/145992/commits/59e131647fe1bdc9a3c53b8643421629a44ff8fc let's see if this trigger some ci errors.

File: torch/_inductor/codegen/simd.py (hunk 4 above)
Comment (FindHao) [reply torch/_inductor/codegen/simd.py:2044]: updated in https://github.com/pytorch/pytorch/pull/145992/commits/59e131647fe1bdc9a3c53b8643421629a44ff8fc

File: torch/_inductor/codegen/simd.py (hunk 5 above)
Comment (FindHao) [reply torch/_inductor/codegen/simd.py:2048]: added in https://github.com/pytorch/pytorch/pull/145992/commits/59e131647fe1bdc9a3c53b8643421629a44ff8fc

File: torch/_inductor/dependencies.py (hunk 6 above)
Comment (FindHao) [reply torch/_inductor/dependencies.py:75]: added in https://github.com/pytorch/pytorch/pull/145992/commits/59e131647fe1bdc9a3c53b8643421629a44ff8fc

File: torch/_inductor/dependencies.py (hunk 7 above)
Comment (FindHao) [reply torch/_inductor/dependencies.py]: yes. replace it with try_get_buffer in https://github.com/pytorch/pytorch/pull/145992/commits/abe008be359ee3d7655f4358bf22bd01fe547bdb

**Comment by FindHao** [comment]
> Can `indirect_broadcast` be computed on the fly rather than storing it as new fields for MemoryDep?

In previous discussion, I agree with @eellison that it's better to have an attribution in MemoryDep. In my thoughts, because the dimension symbol has two changes p0->c0->d0 during creating MemoryDep and the scheduler node, we need to store  those replacements dicts and extra node-replacement mapping in MemoryDep and _RecordLoadStoreInner separately if we want to do it in the fly. So, more attributions are going to be added.
Two replacements information are here https://github.com/pytorch/pytorch/blob/c098385cb3e68682c1f1762ff461d3d59b34651d/torch/_inductor/loop_body.py#L385-L395 and https://github.com/pytorch/pytorch/blob/6889b7b801eb878009447b68aebcc520c10e24d5/torch/_inductor/dependencies.py#L483-L505 .

**Comment by eellison** [comment]
reviewing later today - @FindHao would you mind doing a dashboard run ?
